### ISSUE TITLE
OLA+PreMODIT

### ISSUE THREAD

**Issue by HajimeKawahara** [HajimeKawahara/exojax#279]
After some discussion with @erwanp and @dcmvdbekerom, I decided to implement OLA (OverLap and Add) in PreMODIT framework. I expect this feature will solve
- #277 For a wide wavenumber range (or >4GB memory), jax.fft in fp64 ends with a core dump error. OLA+PreMODIT uses a shorter array for a single FFT.
- sub Lorentian cut. 
I plan to start from the implementation of OLA (scipy.signal.oaconvolve equivalent) using jax (overlapadd branch). 

Some notes:
- As a differentiable FIR, I plan to use `lpf`, the differentiable Faddeeva function
- maybe need [Plinth](https://helios-k2.readthedocs.io/en/latest/helios_k/plinth.html)? Not sure.
- any good idea on a tag/name of OLA+PreMODIT? Precomputation of Shape density and OLA based Resampling,  PreSOLAR? etc

**Comment by HajimeKawahara** [comment]
OLA implemented in #283 and #284

**Comment by HajimeKawahara** [comment]
![ola](https://user-images.githubusercontent.com/15956904/189464961-fe88d0ca-d1ff-4915-9a37-85021dd81efb.PNG)

**Comment by dcmvdbekerom** [comment]
Very nice graph!
The size of a mini-batch is larger than L+M-1 though In `scipy` this is called the `block_size`, it is calculated by the Labert w function and is a function of only the filter-size `M`, see the `_calc_oa_lens()` function in `scipy.signal._signaltools`. It is usually between 8 x M  to 16 x M for typically numbers of M, but can be smaller or larger and doesn't have to be an integer multiple of M.

**Comment by HajimeKawahara** [comment]
Ah, I see. We can directly use `_calc_oa_lens()` to determine the optimal size of  `block_size` because it does not need to be JITtable. I'll try to use it. Thanks!

**Comment by HajimeKawahara** [comment]
You're right. L is much larger than I thought. Assuming the filter length of 50cm-1 @5000cm-1, the optimal block size is ~ 500cm-1. So, for instance, when we would compute the opacity in 1.1-1.4 um range, OLA takes advantages, but 1.1-1.2um is not so much?

  
![optimal_block_size](https://user-images.githubusercontent.com/15956904/189507983-a539f19f-97c7-4f1f-8c58-e5be47f5674f.png)

**Comment by HajimeKawahara** [comment]
![ola](https://user-images.githubusercontent.com/15956904/189552441-828c0c79-8b44-4060-944d-6fae29d5345f.PNG)

|  scipy  |  ExoJAX  |
| ---- | ---- |
| s1 | n L |
|  in1_step |  L  |
|  in2_step = s2  |  M  |
| block_size | L + M -1 |

- The signal should be divided into segments of the same length, L, because of parallelization. 

[code Ref](https://github.com/scipy/scipy/blob/18c6ac77b04ffdcbe21a17c211fe17ce4ece9d68/scipy/signal/_signaltools.py#L774)
```python
in1_step = block_size-s2+1
in2_step = s2
```

**Comment by HajimeKawahara** [comment]
Because I almost finished the implementation of PreMODIT, I restarted PreSOLAR algorithm, which aims to reduce the computational cost rather than device memory use. Also, it includes the cutoff of the tail. 
- [ ] OLA 
- [ ] components tests (single res)
- [ ] xsection (single res)
- [ ] spectrum (single res)
- [ ] components tests for multi res
- [ ] xs for multi res
- [ ] spectrum for multi res

**Comment by HajimeKawahara** [comment]
Breakdown of components tests
- [x] olaform of lbd, nu_grid, shapefilter
- [ ] unbiased hat(lbd),  hat(lbd) means the reshaped and padding array of lbd 
- [ ] differentiable shape filter (i.e. the analytic Voigt used in Direct LPF)
- [ ] tests for olaconv using the above hat(quantities)

**Comment by HajimeKawahara** [comment]
See #365

**Comment by HajimeKawahara** [comment]
implemented in #581
